File: app/frontend/src/api/kanban.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import apiClient from './client'

export interface KanbanBoard {
  id: string
  name: string
  description: string | null
  color: string
  is_archived: boolean
  is_personal: boolean
  created_by: string | null
  created_at: string
  updated_at: string
  task_count: number
  column_count: number
}

export interface KanbanColumn {
  id: string
  board_id: string
  name: string
  name_key?: string | null
  color: string
  is_terminal: boolean
  sort_order: number
}

export interface TaskUser {
  id: string
  full_name: string
  email: string
}

export interface TaskLabel {
  id: string
  name: string
  color: string
  created_by: string | null
  created_at: string
}

export interface Task {
  id: string
  title: string
  description: string | null
  column_id: string
  column: KanbanColumn
  created_by: string
  creator: TaskUser
  assignee_id: string | null
  assignee: TaskUser | null
  priority: 'low' | 'medium' | 'high' | 'critical'
  due_date: string | null
  start_date: string | null
  jira_ticket: string | null
  jira_status: string | null
  jira_status_updated_at: string | null
  sort_order: number
  is_archived: boolean
  created_at: string
  updated_at: string
  labels: TaskLabel[]
}

export interface Attachment {
  id: string
  task_id: string
  original_filename: string
  file_size: number
  mime_type: string
  uploaded_by: string | null
  created_at: string
}

export interface ActivityEntry {
  id: string
  task_id: string
  task_title: string
  action: string
  changes: { field: string; old: string | null; new: string | null }[] | null
  actor: { id: string; full_name: string } | null
  created_at: string
}

export interface ActivityFeedResponse {
  items: ActivityEntry[]
  total: number
  skip: number
  limit: number
}

export interface TaskListResponse {
  items: Task[]
  total: number
  skip: number
  limit: number
}

export const kanbanKeys = {
  all: ['kanban'] as const,
  boards: () => [...kanbanKeys.all, 'boards'] as const,
  board: (id: string) => [...kanbanKeys.all, 'board', id] as const,
  columns: (boardId?: string) => boardId ? [...kanbanKeys.all, 'columns', boardId] as const : [...kanbanKeys.all, 'columns'] as const,
  tasks: (params?: object) => [...kanbanKeys.all, 'tasks', params] as const,
  task: (id: string) => [...kanbanKeys.all, 'task', id] as const,
  labels: () => [...kanbanKeys.all, 'labels'] as const,
}

// ─── Board hooks ──────────────────────────────────────────────────────────────

export function useBoards(params?: { include_archived?: boolean; personal_owner_id?: string }, enabled = true) {
  return useQuery({
    queryKey: [...kanbanKeys.boards(), params],
    queryFn: () => apiClient.get<KanbanBoard[]>('/kanban/boards', { params }).then((r) => r.data),
    enabled,
  })
}

export function useBoard(boardId: string | undefined) {
  return useQuery({
    queryKey: kanbanKeys.board(boardId ?? ''),
    queryFn: () => apiClient.get<KanbanBoard>(`/kanban/boards/${boardId}`).then((r) => r.data),
    enabled: !!boardId,
  })
}

export function useCreateBoard() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: { name: string; description?: string | null; color?: string }) =>
      apiClient.post<KanbanBoard>('/kanban/boards', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.boards() }),
  })
}

export function useUpdateBoard() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...data }: { id: string; name?: string; description?: string | null; color?: string; is_archived?: boolean }) =>
      apiClient.patch<KanbanBoard>(`/kanban/boards/${id}`, data).then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: kanbanKeys.boards() })
      qc.invalidateQueries({ queryKey: kanbanKeys.all })
    },
  })
}

export function useDeleteBoard() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.delete(`/kanban/boards/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.boards() }),
  })
}

export function useColumns(boardId?: string) {
  return useQuery({
    queryKey: kanbanKeys.columns(boardId),
    queryFn: () =>
      apiClient.get<KanbanColumn[]>('/kanban/columns', { params: boardId ? { board_id: boardId } : undefined }).then((r) => r.data),
  })
}

export function useTasks(params?: {
  assignee_id?: string
  team_id?: string
  column_id?: string
  board_id?: string
  priority?: string
  include_archived?: boolean
  search?: string
  label_id?: string
  limit?: number
}) {
  return useQuery({
    queryKey: kanbanKeys.tasks(params),
    queryFn: () =>
      apiClient.get<TaskListResponse>('/kanban/tasks', { params: { limit: 500, ...params } }).then((r) => r.data),
  })
}

export function useCreateColumn() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: { name: string; color: string; is_terminal: boolean; sort_order: number; board_id?: string }) =>
      apiClient.post<KanbanColumn>('/kanban/columns', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.all }),
  })
}

export function useUpdateColumn() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...data }: { id: string; name?: string; color?: string; is_terminal?: boolean; sort_order?: number }) =>
      apiClient.patch<KanbanColumn>(`/kanban/columns/${id}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.all }),
  })
}

export function useDeleteColumn() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.delete(`/kanban/columns/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.all }),
  })
}

export function useReorderColumns() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (items: { id: string; sort_order: number }[]) =>
      apiClient.put<KanbanColumn[]>('/kanban/columns/reorder', items).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.columns() }),
  })
}

export function useCreateTask() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: {
      title: string
      column_id: string
      description?: string
      assignee_id?: string
      priority?: string
      due_date?: string
      start_date?: string
      jira_ticket?: string
      label_ids?: string[]
    }) => apiClient.post<Task>('/kanban/tasks', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.all }),
  })
}

export function useUpdateTask() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...data }: {
      id: string
      title?: string
      description?: string
      assignee_id?: string | null
      priority?: string
      due_date?: string | null
      start_date?: string | null
      jira_ticket?: string | null
      is_archived?: boolean
      label_ids?: string[] | null
    }) => apiClient.patch<Task>(`/kanban/tasks/${id}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.all }),
  })
}

export function useMoveTask() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, column_id, sort_order }: { id: string; column_id: string; sort_order: number }) =>
      apiClient.patch<Task>(`/kanban/tasks/${id}/move`, { column_id, sort_order }).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.all }),
  })
}

export function useDeleteTask() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.delete(`/kanban/tasks/${id}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.all }),
  })
}

export interface TaskComment {
  id: string
  task_id: string
  user_id: string | null
  author: TaskUser | null
  content: string
  created_at: string
  updated_at: string
}

export const commentKeys = {
  list: (taskId: string) => [...kanbanKeys.all, 'comments', taskId] as const,
}

export function useTaskComments(taskId: string | undefined) {
  return useQuery({
    queryKey: commentKeys.list(taskId ?? ''),
    queryFn: () =>
      apiClient.get<TaskComment[]>(`/kanban/tasks/${taskId}/comments`).then((r) => r.data),
    enabled: !!taskId,
  })
}

export function useCreateComment(taskId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (content: string) =>
      apiClient.post<TaskComment>(`/kanban/tasks/${taskId}/comments`, { content }).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: commentKeys.list(taskId) }),
  })
}

export function useDeleteComment(taskId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (commentId: string) =>
      apiClient.delete(`/kanban/tasks/${taskId}/comments/${commentId}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: commentKeys.list(taskId) }),
  })
}

export interface TaskHistoryChange {
  field: string
  old: string | null
  new: string | null
}

export interface TaskHistoryEntry {
  id: string
  task_id: string
  action: string
  changes: TaskHistoryChange[] | null
  actor: TaskUser | null
  created_at: string
}

export const historyKeys = {
  list: (taskId: string) => [...kanbanKeys.all, 'history', taskId] as const,
}

export function useTaskHistory(taskId: string | undefined) {
  return useQuery({
    queryKey: historyKeys.list(taskId ?? ''),
    queryFn: () =>
      apiClient.get<TaskHistoryEntry[]>(`/kanban/tasks/${taskId}/history`).then((r) => r.data),
    enabled: !!taskId,
  })
}

// ─── Subtasks ────────────────────────────────────────────────────────────────

export interface Subtask {
  id: string
  task_id: string
  title: string
  is_completed: boolean
  sort_order: number
  created_at: string
}

export const subtaskKeys = {
  list: (taskId: string) => [...kanbanKeys.all, 'subtasks', taskId] as const,
}

export function useTaskSubtasks(taskId: string | null | undefined) {
  return useQuery({
    queryKey: subtaskKeys.list(taskId ?? ''),
    queryFn: () =>
      apiClient.get<Subtask[]>(`/kanban/tasks/${taskId}/subtasks`).then((r) => r.data),
    enabled: !!taskId,
  })
}

export function useCreateSubtask(taskId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: { title: string; sort_order?: number }) =>
      apiClient.post<Subtask>(`/kanban/tasks/${taskId}/subtasks`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: subtaskKeys.list(taskId) }),
  })
}

export function useUpdateSubtask(taskId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ subtaskId, ...data }: { subtaskId: string; title?: string; is_completed?: boolean; sort_order?: number }) =>
      apiClient.patch<Subtask>(`/kanban/tasks/${taskId}/subtasks/${subtaskId}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: subtaskKeys.list(taskId) }),
  })
}

export function useDeleteSubtask(taskId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (subtaskId: string) =>
      apiClient.delete(`/kanban/tasks/${taskId}/subtasks/${subtaskId}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: subtaskKeys.list(taskId) }),
  })
}

// ─── Bulk Operations ─────────────────────────────────────────────────────────

export function useBulkUpdateTasks() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: {
      task_ids: string[]
      column_id?: string
      assignee_id?: string | null
      priority?: string
      is_archived?: boolean
    }) => apiClient.patch<{ updated: number }>('/kanban/tasks/bulk', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.all }),
  })
}

// ─── Attachments ─────────────────────────────────────────────────────────────

export const attachmentKeys = {
  list: (taskId: string) => [...kanbanKeys.all, 'attachments', taskId] as const,
}

export function useTaskAttachments(taskId: string | null | undefined) {
  return useQuery({
    queryKey: attachmentKeys.list(taskId ?? ''),
    queryFn: () =>
      apiClient.get<Attachment[]>(`/kanban/tasks/${taskId}/attachments`).then((r) => r.data),
    enabled: !!taskId,
  })
}

export function useUploadAttachment(taskId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (file: File) => {
      const fd = new FormData()
      fd.append('file', file)
      return apiClient.post<Attachment>(`/kanban/tasks/${taskId}/attachments`, fd, {
        headers: { 'Content-Type': 'multipart/form-data' },
      }).then((r) => r.data)
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: attachmentKeys.list(taskId) }),
  })
}

export function useDeleteAttachment(taskId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (attId: string) =>
      apiClient.delete(`/kanban/tasks/${taskId}/attachments/${attId}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: attachmentKeys.list(taskId) }),
  })
}

export function getAttachmentDownloadUrl(taskId: string, attId: string) {
  return `/api/v1/kanban/tasks/${taskId}/attachments/${attId}/download`
}

// ─── Labels ───────────────────────────────────────────────────────────────────

export function useLabels() {
  return useQuery({
    queryKey: kanbanKeys.labels(),
    queryFn: () => apiClient.get<TaskLabel[]>('/kanban/labels').then((r) => r.data),
  })
}

export function useCreateLabel() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (data: { name: string; color: string }) =>
      apiClient.post<TaskLabel>('/kanban/labels', data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.labels() }),
  })
}

export function useUpdateLabel() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...data }: { id: string; name?: string; color?: string }) =>
      apiClient.patch<TaskLabel>(`/kanban/labels/${id}`, data).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: kanbanKeys.labels() }),
  })
}

export function useDeleteLabel() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.delete(`/kanban/labels/${id}`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: kanbanKeys.labels() })
      qc.invalidateQueries({ queryKey: kanbanKeys.all }) // refresh tasks that had this label
    },
  })
}

// ─── Activity Feed ────────────────────────────────────────────────────────────

export const activityKeys = {
  feed: (p: object) => [...kanbanKeys.all, 'activity', p] as const,
}

export function useActivityFeed(params: { skip?: number; limit?: number } = {}) {
  return useQuery({
    queryKey: activityKeys.feed(params),
    queryFn: () =>
      apiClient.get<ActivityFeedResponse>('/kanban/activity', { params }).then((r) => r.data),
  })
}

File: app/backend/app/schemas/kanban.py
import re
import uuid
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field, field_validator

_JIRA_RE = re.compile(r'^[A-Za-z][A-Za-z0-9_]*-\d+$')


# ─── Label schemas ────────────────────────────────────────────────────────────

class LabelCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    color: str = "#6366f1"

    @field_validator("color")
    @classmethod
    def validate_color(cls, v: str) -> str:
        if not v.startswith("#") or len(v) not in (4, 7):
            raise ValueError("Renk geçerli bir HEX değeri olmalıdır.")
        return v


class LabelUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    color: Optional[str] = None


class LabelResponse(BaseModel):
    id: uuid.UUID
    name: str
    color: str
    created_by: Optional[uuid.UUID]
    created_at: datetime
    model_config = {"from_attributes": True}


# ─── Board schemas ────────────────────────────────────────────────────────────

class BoardCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=1000)
    color: str = "#6366f1"

    @field_validator("color")
    @classmethod
    def validate_color(cls, v: str) -> str:
        if not v.startswith("#") or len(v) not in (4, 7):
            raise ValueError("Renk geçerli bir HEX değeri olmalıdır.")
        return v


class BoardUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=1000)
    color: Optional[str] = None
    is_archived: Optional[bool] = None


class BoardResponse(BaseModel):
    id: uuid.UUID
    name: str
    description: Optional[str]
    color: str
    is_archived: bool
    is_personal: bool = False
    created_by: Optional[uuid.UUID]
    created_at: datetime
    updated_at: datetime
    task_count: int = 0
    column_count: int = 0
    model_config = {"from_attributes": True}


# ─── Column schemas ───────────────────────────────────────────────────────────

class ColumnCreate(BaseModel):
    name: str
    color: str = "#e2e8f0"
    is_terminal: bool = False
    sort_order: int = 0
    board_id: Optional[uuid.UUID] = None

    @field_validator("color")
    @classmethod
    def validate_color(cls, v: str) -> str:
        if not v.startswith("#") or len(v) not in (4, 7):
            raise ValueError("Renk geçerli bir HEX değeri olmalıdır.")
        return v


class ColumnUpdate(BaseModel):
    name: Optional[str] = None
    color: Optional[str] = None
    is_terminal: Optional[bool] = None
    sort_order: Optional[int] = None


class ColumnReorderItem(BaseModel):
    id: uuid.UUID
    sort_order: int


class ColumnResponse(BaseModel):
    id: uuid.UUID
    board_id: uuid.UUID
    name: str
    name_key: Optional[str] = None
    color: str
    is_terminal: bool
    sort_order: int
    model_config = {"from_attributes": True}


class UserBasic(BaseModel):
    id: uuid.UUID
    full_name: str
    email: str
    model_config = {"from_attributes": True}


class TaskCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=500)
    column_id: uuid.UUID
    description: Optional[str] = Field(None, max_length=50000)
    assignee_id: Optional[uuid.UUID] = None
    priority: str = "medium"
    due_date: Optional[date] = None
    start_date: Optional[date] = None
    jira_ticket: Optional[str] = Field(None, max_length=50)
    label_ids: list[uuid.UUID] = Field(default_factory=list)

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: str) -> str:
        if v not in ("low", "medium", "high", "critical"):
            raise ValueError("Öncelik low, medium, high veya critical olmalıdır.")
        return v

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: str) -> str:
        v = v.strip()
        if len(v) < 2:
            raise ValueError("Başlık en az 2 karakter olmalıdır.")
        return v

    @field_validator("jira_ticket")
    @classmethod
    def validate_jira_ticket(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            v = v.strip()
            if v and not _JIRA_RE.match(v):
                raise ValueError("Jira ticket formatı geçersiz (örn: PROJ-123).")
        return v or None


class TaskUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=2, max_length=500)
    description: Optional[str] = Field(None, max_length=50000)
    assignee_id: Optional[uuid.UUID] = None
    priority: Optional[str] = None
    due_date: Optional[date] = None
    start_date: Optional[date] = None
    jira_ticket: Optional[str] = Field(None, max_length=50)
    is_archived: Optional[bool] = None
    label_ids: Optional[list[uuid.UUID]] = None  # None = don't change; [] = clear all

    @field_validator("jira_ticket")
    @classmethod
    def validate_jira_ticket(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            v = v.strip()
            if v and not _JIRA_RE.match(v):
                raise ValueError("Jira ticket formatı geçersiz (örn: PROJ-123).")
        return v or None


class TaskMoveRequest(BaseModel):
    column_id: uuid.UUID
    sort_order: int


class TaskResponse(BaseModel):
    id: uuid.UUID
    title: str
    description: Optional[str]
    column_id: uuid.UUID
    column: ColumnResponse
    created_by: uuid.UUID
    creator: UserBasic
    assignee_id: Optional[uuid.UUID]
    assignee: Optional[UserBasic]
    priority: str
    due_date: Optional[date]
    start_date: Optional[date] = None
    jira_ticket: Optional[str]
    jira_status: Optional[str]
    jira_status_updated_at: Optional[datetime]
    sort_order: int
    is_archived: bool
    created_at: datetime
    updated_at: datetime
    labels: list[LabelResponse] = Field(default_factory=list)
    model_config = {"from_attributes": True}


class TaskListResponse(BaseModel):
    items: list[TaskResponse]
    total: int
    skip: int
    limit: int


class TaskCommentCreate(BaseModel):
    content: str

    @field_validator("content")
    @classmethod
    def validate_content(cls, v: str) -> str:
        v = v.strip()
        if len(v) < 1:
            raise ValueError("Yorum boş olamaz.")
        if len(v) > 2000:
            raise ValueError("Yorum en fazla 2000 karakter olabilir.")
        return v


class TaskCommentResponse(BaseModel):
    id: uuid.UUID
    task_id: uuid.UUID
    user_id: Optional[uuid.UUID]
    author: Optional[UserBasic]
    content: str
    created_at: datetime
    model_config = {"from_attributes": True}


class TaskHistoryChange(BaseModel):
    field: str
    old: Optional[str]
    new: Optional[str]


class TaskHistoryEntry(BaseModel):
    id: uuid.UUID
    task_id: uuid.UUID
    action: str
    changes: Optional[list[TaskHistoryChange]]
    actor: Optional[UserBasic]
    created_at: datetime
    model_config = {"from_attributes": True}


# ─── Subtask schemas ──────────────────────────────────────────────────────────

class SubtaskCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    sort_order: int = 0


class SubtaskUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    is_completed: Optional[bool] = None
    sort_order: Optional[int] = None


class SubtaskResponse(BaseModel):
    id: uuid.UUID
    task_id: uuid.UUID
    title: str
    is_completed: bool
    sort_order: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class AttachmentResponse(BaseModel):
    id: uuid.UUID
    task_id: uuid.UUID
    original_filename: str
    file_size: int
    mime_type: str
    uploaded_by: Optional[uuid.UUID]
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)
